app/gui/src/providers/openedProjects.ts

```diff
@@ -0,0 +1,84 @@
+import { createContextStore } from '@/providers'
+import { WidgetRegistry } from '@/providers/widgetRegistry'
+import { createGraphStore, GraphStore } from '@/stores/graph'
+import { createProjectStore, LsUrls, ProjectStore } from '@/stores/project'
+import { createProjectNameStore, ProjectNameStore } from '@/stores/projectNames'
+import { createSuggestionDbStore, SuggestionDbStore } from '@/stores/suggestionDatabase'
+import { ToValue } from '@/util/reactivity'
+import { EffectScope, effectScope, shallowReactive } from 'vue'
+
+/** All stores of a single opened project */
+export interface OpenedProject {
+  store: ProjectStore
+  names: ProjectNameStore
+  suggestionDb: SuggestionDbStore
+  graph: GraphStore
+  widgetRegistry: WidgetRegistry
+}
+
+/**
+ * Properties of the project.
+ *
+ * This is a subset of ProjectView props which is used to set up the store.
+ */
+export interface ProjectProps {
+  projectId: string
+  projectNamespace: ToValue<string | undefined>
+  projectInitialName: string
+  projectDisplayedName: ToValue<string>
+  renameProject: (newName: string) => void
+  engine: LsUrls
+}
+
+/**
+ * A type for Opened Project Store.
+ */
+export type OpenedProjectsStore = ReturnType<typeof injectOpenedProjects>
+
+/**
+ * Opened Projects Store
+ *
+ * This store maintains all "substores" of opened projects. When an opened project registers,
+ * the names, project, suggestionDb, graph and widgetRegistry stores are created and available
+ * through `get` method.
+ *
+ * See also `WithCurrentProject` component which allows setting one opened project as "default"
+ * for component subtree.
+ */
+export const [provideOpenedProjects, injectOpenedProjects] = createContextStore(
+  'opened-projects',
+  () => {
+    const projects = shallowReactive(
+      new Map<string, OpenedProject & { storesScope: EffectScope }>(),
+    )
+
+    function registerProject(props: ProjectProps) {
+      const { projectId } = props
+      const storesScope = effectScope()
+
```

Comment:
There should be an `assert` checking if we aren't registering an already registered project.